locker.go

```diff
@@ -0,0 +1,344 @@
+package consul
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// A Locker exposes methods for acquiring locks on keys of the consul key/value
+// store.
+type Locker struct {
+	// The client used to send requests to the consul agent. If nil, the default
+	// client is used instead.
+	Client *Client
+
+	// LockDelay is the amount of time that a lock will stay held if it hasn't
+	// been released and the session that was attached to it expired.
+	LockDelay time.Duration
+}
+
+// Lock acquires locks on the given keys. The method blocks until the locks were
+// acquired, or the context was camceled. The returned context will be canceled
+// when the locks are released (by calling the cancellation function), or if the
+// ownership was lost, in this case the contexts' Err method returns Unlocked.
+func (l *Locker) Lock(ctx context.Context, keys ...string) (context.Context, context.CancelFunc) {
+	retryInterval := 1 * time.Second
+	sessionCtx, sessionCancel := l.withSession(ctx, "lock: %v", keys)
+
+	deadline, ok := ctx.Deadline()
+	if ok {
+		retryInterval = deadline.Sub(time.Now()) / 10
+	}
+
+	locks := make([]ctxCancel, 0, len(keys))
+	sortedKeys := sortedKeys(keys)
+
+	for {
+		for _, key := range sortedKeys {
+			lockCtx, lockCancel, _ := l.tryLock(sessionCtx, key)
+			if lockCtx == nil {
+				break
+			}
+			locks = append(locks, ctxCancel{lockCtx, lockCancel})
+		}
+
+		if len(locks) == len(keys) {
+			multi := newMultiCtx(sortedKeys, locks)
+			return multi, func() { multi.cancel(); sessionCancel() }
+		}
+
+		for _, lock := range locks {
+			lock.cancel()
+		}
+		locks = locks[:0]
+
+		timer := time.NewTicker(retryInterval)
+		select {
+		case <-timer.C:
+			timer.Stop()
+		case <-ctx.Done():
+			timer.Stop()
+			sessionCancel()
+			return errorContext(ctx, ctx.Err())
+		}
+	}
+}
+
+// TryLock attempts to acquire a lock on one of the given keys. The returned
+// context will be canceled when the lock is released (by calling the
+// cancellation function), or if the ownership was lost, in this case the
+// context's Err method returns Unlocked.
+// The method never blocks, if it fails to acquire any of the locks it returns
+// a canceled context.
+func (l *Locker) TryLock(ctx context.Context, keys ...string) (context.Context, context.CancelFunc) {
```

Comment:
I thought about it as I was writing the documentation, it could be a bit confusing, some alternative names I thought of:
- TryLockOne
- TryLockAny